tools/dataflow-throttling/README.md

```diff
@@ -0,0 +1,111 @@
+
+This is a generic library intended to implement client-side throttling by rejecting requests that have a high probability of being rejected locally on Dataflow nodes. Irrespective of Batch or Streaming, when Dataflow pipeline is sending HTTP requests to the server with input elements as payload, once the request has been processed by the backend, it should send a response back to the pipeline whether the http request has been accepted or rejected. The error code depends up on the backend throttling implementation (defaults to HTTP code 429).
+
+![DataflowThrottling DAG](img/dataflow-throttling-dag.png "Dataflow Throttling DAG")
+
+### Throttle with multiple groups
+
+Implemented DynamicThrottlingTransform using stateful processing with single group of elements that represent requests to the external service. We’re going to maintain the following states.
+
+* incomingRequests - requests stored in Apache Beam bag state
+* acceptedRequests number
+* totalRequestsProcessed number
+
+The count of the incoming requests and accepted requests should be equal under normal conditions. Once the requests start getting rejected, the number of processing requests gets decreased by the difference of incoming requests and accepted requests.
+
+Pipeline will process the payload in multiple number of groups as well. To achieve this, pipeline will transforms the single PCollection[Bounded/Unbounded] into multiple groups.
+
+* Converts the Input PCollection<<T>T</T>> into PCollection<<T>Key,Value</T>>. Here, Key will be randomly assigned integer and Value will be payload.
+* All the elements with the same key will be grouped together and emit another PCollection<<T>Key,Iterable</T>>>, where key will be unique and Iterable<<T>T</T>> will be the collection of elements with the same key.
+* Adaptive throttling will be applied to each group[State cell] accordingly.
+* Pipeline will process the each state cell using timely processing. That said each state cell will be processed after reaching a predefined interval of  time. A user function will be applied on a group of elements.
+* clientCall is a user defined function where the events will be sent to the backend node, Applies to each event in the pipeline. Based on the response of the clientCall respective counters will get incremented.
+* The variables which stores the number of requested and accepted requests are zeroed out after a certain interval of time (defaults to 1 min). This reset speeds up client recovery if server throughput was limited due to limitations on the server side not caused by the client.
+
+### Adaptive throttling
+
+Dataflow pipeline will maintain the number of requests it has sent to the backend [total_requests], the number of requests got accepted by the backend [total_accepts]. Using stateTimer, It will process a batch of elements from each state value and removes the processed elements from the state value. Request rejection probability will be calculated as follows.
```

Comment:
"will maintain" -> maintains
, the number must be "and the numbebr"
It -> it
removes -> remove